dataclients/kubernetes/crd.md

```diff
@@ -0,0 +1,1032 @@
+    - oauthTokeninfoAllScope("myapp.read")
+```
+
+#### Complex routes with ratelimits based on tokens
+
+A complex route case could specify different ratelimits for POST and PUT
+requests to /api/resource for clients.  Clients that have JWT/OAuth2
+Tokens from issuer https://accounts.google.com with email
+"important@example.org" get 20/equest per minute, other clients with
+Token with issuer https://accounts.google.com or
+https://accounts.github.com get only 2.
+
+Additionally for all other HTTP methods ratelimit for each client 10
+requests per hour based on "Authorization" header and make sure the
+Token is valid and has the scopes to read and list the resource.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+metadata:
+  name: api
+spec:
+  hosts:
+  - api.example.org
+  - api.service.example.net
+  backends:
+  - name: api-svc
+    type: service
+    serviceName: api-service-v1
+    servicePort: 80
+  defaultBackends:
+  - backendName: api-svc
+  routes:
+  - path: /api/resource
+    methods:
+    - post
+    - put
+    filters:
+    - ratelimit(20, "1m")
+    - oauthTokeninfoAllKV("iss", "https://accounts.google.com", "email", "important@example.org")
+    predicates:
+    - JWTPayloadAllKV("iss", "https://accounts.google.com", "email", "skipper-router@googlegroups.com")
+  - path: /api/resource
+    methods:
+    - post
+    - put
+    filters:
+    - ratelimit(2, "1m")
+    - oauthTokeninfoAnyKV("iss", "https://accounts.google.com", "iss", "https://accounts.github.com")
+  - path: /api/resource
+    filters:
+    - clientRatelimit(10, "1h", "Authorization")
+    - oauthTokeninfoAllScope("read.resource", "list.resource")
+```
+
+#### Traffic switching
+
+If we want to traffic switch, we want to make sure all path endpoints
+we have configured are switched at the same time and get the same
+traffic split applied. To not be ambiguous with multiple backends, we
+need to set the `weight` and both backends have to be
+`defaultBackends`, if we want to traffic switch.
+
+```yaml
+apiVersion: zalando.org/v1
+kind: RouteGroup
+metadata:
+  name: my-routes
+spec:
+  hosts:
+  - api.example.org
+  - api.service.example.net
+  backends:
+  - name: api-svc
+    type: service
+    serviceName: api-service-v1
+    servicePort: 80
+  - name: api-svc-v2
+    type: service
+    serviceName: foo-service-v2
+    servicePort: 80
+  defaultBackends:
+  - backendName: api-svc
+    weight: 80
+  - backendName: api-svc-v2
+    weight: 20
+  routes:
+  - path: /api/resource
+    filters:
+    - ratelimit(200, "1m")
+    - oauthTokeninfoAllKV("iss", "https://accounts.google.com", "email", "skipper-router@googlegroups.com")
+    predicates:
+    - JWTPayloadAllKV("iss", "https://accounts.google.com", "email", "skipper-router@googlegroups.com")
+  - path: /api/resource
+    filters:
+    - ratelimit(20, "1m")
+    - oauthTokeninfoAllKV("iss", "https://accounts.google.com")
+```
+
+#### A/B test
+
+A/B test via cookie `canary`, used for sticky sessions.
```

Comment:
this is probably just a naming issue: we should try not to mix A/B tests and canary tests because they are quite different high level concepts. We use canary tests to make sure that the new versions of an application are stable. A/B tests may span over multiple versions of an application, e.g. when a variant or each variants need a bugfix, but we want to continue with the A/B test.

Reply:
Feel free to change it, but IIRC I used canary, because it was actually testing that the B branch works, before switching 10% customers to it.

Reply:
I think 'canary' is the better name. Replacing 'A/B' with 'canary'.

Reply:
Actually... not in every case. Ignore the above.